receiver/prometheusreceiver/metrics_receiver.go

```diff
@@ -51,62 +52,55 @@ func newPrometheusReceiver(logger *zap.Logger, cfg *Config, next consumer.Metric
 
 // Start is the method that starts Prometheus scraping and it
 // is controlled by having previously defined a Configuration using perhaps New.
-func (pr *pReceiver) Start(_ context.Context, host component.Host) error {
-	pr.startOnce.Do(func() {
-		ctx := context.Background()
-		c, cancel := context.WithCancel(ctx)
-		pr.cancel = cancel
-		c = obsreport.ReceiverContext(c, pr.cfg.Name(), "http", pr.cfg.Name())
-		var jobsMap *internal.JobsMap
-		if !pr.cfg.UseStartTimeMetric {
-			jobsMap = internal.NewJobsMap(2 * time.Minute)
+func (r *pReceiver) Start(ctx context.Context, host component.Host) error {
+	logger := internal.NewZapToGokitLogAdapter(r.logger)
+
+	var startErr error
+	r.startOnce.Do(func() {
+		discoveryCtx := context.Background()
+		discoveryCtx, cancel := context.WithCancel(ctx)
+		r.cancelFunc = cancel
```

Comment:
Line 60 can be removed, the `discoveryCtx` is always re-written correct? (but see the other comment about why the discoverCtx should not inherit from the context passed to start)

Reply:
This is very interesting pattern that I observed in a lot of places (maybe you can help me understand better how can we avoid this "possible" mistake).

Extending the incoming context from the Start func I think is wrong, when the context is saved for "async" work. In this case you pass `discoveryCtx` to the discovery manager, and save a reference to it, then check if the context is cancelled to understand when to stop.

Here is why I think it is wrong:
* The caller of the "Start" will possibly add a "deadline" to the context to ensure that the components start within a certain amount of time. Or can define its own cancelation to the context (for whatever reasons).
* In case of extending from that context, you inherit the cancellation. And from the deadline example the extended context will be cancelled (the `WithTimeout` suggest to call the cancellation func when done), the caller is consider done when you return from `Start`, so the extended context `discoveryCtx` will be closed almost immediately after the Start ends.

I usually try to tell users that when you schedule "async" work you cannot inherit from "sync" context because the caller may cancel the context when you return (or you need to do a trick like I did here https://github.com/open-telemetry/opentelemetry-collector/blob/add864b84b65796451ff0e90875cdfd34ca21f32/exporter/exporterhelper/queued_retry.go#L277 where you inherit everything but the cancellation).

This is a long comment, hope you understand the problem, we had this https://github.com/open-telemetry/opentelemetry-collector/issues/1813 where exactly this happened during the execution of an gRPC request we were scheduling async work in the queue_retry and return, the gRPC when we return will call done on the timeout return func, and the saved context will be "cancelled" so all the export calls will fail because we will pass a done context.

Reply:
I'm not familiar with the lifecycle of the components so, I looked at the the Host interface and it recommended me to start a new background context:
```
	// If the component needs to perform a long-running starting operation then it is recommended
	// that Start() returns quickly and the long-running operation is performed in background.
	// In that case make sure that the long-running operation does not use the context passed
	// to Start() function since that context will be cancelled soon and can abort the long-running operation.
	// Create a new context from the context.Background() for long-running operations.
	Start(ctx context.Context, host Host) error
```

I'm not familiar with the lifecycle of the components. Do we expect components to handle the cancellation/timeout signals in Start? If yes, we should mentioned it in the interface. If you clarify what the components should do with the context, I can try to make modifications. Unfortunately the Prometheus APIs doesn't allow us to pass different contexts to discovery.NewManager and discoveryManager.Run and that's the real problem here.

In an ideal situations, this code should be written like the following but the Prometheus APIs doesn't allow us to do it:

```
func (r *pReceiver) Start(ctx context.Context, host component.Host) error {
discoveryManager := discovery.NewManager(ctx, logger)

// ...

discoveryRunCtx := context.Background()
discoveryRunCtx, cancel := context.WithCancel(ctx)
discoveryManager.Run(discoveryRunCtx);
```

Splitting the cancellation and timeout sounds like a hack to get around the limitations of the APIs. Am I right?

Reply:
Btw, I removed the `discoveryCtx := context.Background()` because it seems like the only way ctx is cancelled when things are timed out or something wrong happened in the starting process. It's ok to cancel the discovery manager in both of those cases. Let me know if this looks ok.

Reply:
@rakyll I think code will be simpler:

```golang
// In the service try to start all components in maximum 30 seconds
func (srv *Service) startAllComponents() {
  ctx, cancelFunc := context.WithTimeout(context.Background(), 30*time.Second)
  defer cancelFunc()
  for i := range srv.components {
    srv.components[i].Start(ctx, srv.host)
  }
}


func (r *pReceiver) Start(ctx context.Context, host component.Host) error {
  discoveryCtx, cancel := context.WithCancel(ctx)
  r.cancelFunc = cancel
  // Manager will run until discoveryCtx is cancelled.
  discoveryManager := discovery.NewManager(discoveryCtx, logger)
  ...
}
```

Problems:
1. As you can see the `discoveryCtx` will be cancelled when the `startAllComponents` ends, way before the service shutting down.

Reply:
> In an ideal situation, this code should be written like the following but the Prometheus APIs doesn't allow us to do it:

```
func (r *pReceiver) Start(ctx context.Context, host component.Host) error {
discoveryManager := discovery.NewManager(ctx, logger)

// ...

discoveryRunCtx := context.Background()
discoveryRunCtx, cancel := context.WithCancel(ctx)
discoveryManager.Run(discoveryRunCtx);
}
```

Yes, the context pass to run should be decoupled from the context pass to Start. You are right.

Reply:
Switched back to the background context.